pkg/loghttp/v1/query.go

```diff
@@ -0,0 +1,197 @@
+package v1
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+//QueryStatus
+type QueryStatus string
+
+//QueryStatus values
+const (
+	QueryStatusSuccess = "success"
+)
+
+//QueryResponse represents the http json response to a label query
+type QueryResponse struct {
+	Status string            `json:"status"`
+	Data   QueryResponseData `json:"data"`
+}
+
+//ResultType
+type ResultType string
+
+//ResultType values
+const (
+	ResultTypeStream = "streams"
+	ResultTypeVector = "vector"
+	ResultTypeMatrix = "matrix"
+)
+
+type ResultValue interface {
+	Type() ResultType
+}
+
+//QueryResponseData represents the http json response to a label query
+type QueryResponseData struct {
+	ResultType ResultType  `json:"resultType"`
+	Result     ResultValue `json:"result"`
+}
+
+func (Streams) Type() ResultType { return ResultTypeStream }
+func (Vector) Type() ResultType  { return ResultTypeVector }
+func (Matrix) Type() ResultType  { return ResultTypeMatrix }
+
+type Streams []Stream
+
+//Stream represents a log stream.  It includes a set of log entries and their labels.
+type Stream struct {
+	Labels  LabelSet `json:"stream"`
+	Entries []Entry  `json:"values"`
+}
+
+//Entry represents a log entry.  It includes a log message and the time it occurred at.
+type Entry struct {
+	Timestamp time.Time
+	Line      string
+}
+
+//MarshalJSON converts an Entry object to be prom compatible for http queries
+func (e *Entry) MarshalJSON() ([]byte, error) {
+	l, err := json.Marshal(e.Line)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(fmt.Sprintf("[\"%d\",%s]", e.Timestamp.UnixNano(), l)), nil
+}
+
+//UnmarshalJSON
+func (e *Entry) UnmarshalJSON(data []byte) error {
+	var unmarshal []string
+
+	err := json.Unmarshal(data, &unmarshal)
+	if err != nil {
+		return err
+	}
+
+	t, err := strconv.ParseInt(unmarshal[0], 10, 64)
+	if err != nil {
+		return err
+	}
+
+	e.Timestamp = time.Unix(0, t)
+	e.Line = unmarshal[1]
+
+	return nil
+}
+
+//Vector
+type Vector []model.Sample
+
+//Matrix
+type Matrix []model.SampleStream
+
+func NewStreams(s logql.Streams) (Streams, error) {
```

Comment:
I kind of feel in instances like these it might make more sense to define these in (for instance) logql? If we're intending loghttp to be the package consumed externally, then we're using what are intended to be internal types in this external package. 

In my opinion this might be nicer since it'll clean up all the functions in this package:

```go
package logql 

// HTTPStreams converts Streams into a publicly consumable type
func (s Streams) HTTPStreams() loghttp.Streams { ... }
```

This is subjective though, feel free to disagree.